Temporary Images Resizing on Resolution Change

authenticate opened the pull request:
Hi,

I uploaded some changes to address to the temporary image resizing issue.  You should be able to change the resolution in the map, shop, and battle modes and the background images will resize appropriately.

I am not sure if this is exactly how you would want to implement this fix.  It is kind of aggressive.  However, I think it illustrates the primary idea of having to iterate up the game state stack; redrawing the game state at each iteration.

In addition, I think there was an implicit conversion causing a memory leak in shop mode.  I uploaded some comments in 'shop.cpp'.  It will probably be easier to look up those comments with the code as context then me trying to explain here.

Let me know what you think and/or if you have any questions or comments.

Thanks.

Bertram25 commented:
I'll be on it soon. Sorry to keep you waiting... :)

Bertram25 commented:
Hi @authenticate 

I've reviewed and tested a bit your changes (finally). First of all thanks for all the style fixes and the use of c++11 override where relevant. :)

> I am not sure if this is exactly how you would want to implement this fix. It is kind of aggressive. However, I think it illustrates the primary idea of having to iterate up the game state stack; redrawing the game state at each iteration.

It is aggressive indeed. The first thing that is striking me is that you prefer redrawing the whole game stack, rather than storing the bg image in a common temp image that can be resized and relocated on screen to fit the current resolution. I do think your fixes on the viewport part already fixed the image size, and rest is indeed easier said than done, of course, ;) but wonder whether the file to a common temp location and reloading it in that case would suffice. IMHO, it will. I must say I do not care if the texture isn't exactly fitting after a resolution change.

The second thing that is striking me is that you reused the reset function, and I wonder if it was the best choice, but it doesn't apply IMHO as I'd prefer having a background texture saved and reloaded if necessary.

> In addition, I think there was an implicit conversion causing a memory leak in shop mode. I uploaded some comments in 'shop.cpp'. It will probably be easier to look up those comments with the code as context then me trying to explain here.

Wow, that one is weird, indeed. Could we talk about that one in a second commit? The first change is IMHO already complex enough. But nice catch in any case! :)
Ah, and could you rebase? There have been a few changes.

Thanks a lot and best regards,

authenticate commented:
I rebased everything and moved the shop mode change into a separate pull request to make things easier to manage.

I do not have much more time tonight.  I will ask some more questions in a couple days or so about how to fix it as it switches to use temporary textures.

I figured some of this would not be how you wanted it implemented.  However, I just wanted to get something implemented and working.  Then, we can revise it and update it.

Have a good one.

authenticate commented:
I pushed up some updates.

I changed 'pause.cpp' to scale the original texture.  Let me know if this approach is alright.  If it is, I will go through and update the other sections of the code to use it and remove some of the unneeded changes.

Thanks again.

Bertram25 commented:
> I figured some of this would not be how you wanted it implemented. However, I just wanted to get something implemented and working. Then, we can revise and update it. :)

Yep, please don't misunderstand my comments as mistrust or anything offensive, I do respect and appreciate your work. :) And it's good to get fresh ideas about such technical issues.
I'll have a look at all that tomorrow whenever I've got some spare-time.

Thanks a lot!

Bertram25 commented:
> I changed 'pause.cpp' to scale the original texture. Let me know if this approach is alright. If it is, I will go through and update the other sections of the code to use it and remove some of the unneeded changes.

I've seen the changes in the pause mode and they look good so far (Still need to test them alone).
Could you break the style fixes and the changes like in the pause mode into two PRs? (Since there is a lot of style fixes.)

authenticate commented:
I updated this pull request to contains the style fixes and very minor changes.

I'll open a new pull request containing the temporary image resolution fix.

Bertram25 commented:
arg, I didn't receive any notifications about your work!!
I will have a look asap.

Bertram25 commented on src/engine/mode_manager.cpp:
```diff
@@ -262,12 +262,20 @@ GameMode *ModeEngine::GetTop()
 
+size_t ModeEngine::Size() const
+{
+    return _game_stack.size();
```
Do we still need that change about the game mode manager?

authenticate replied:
I'll remove the Size function.

The change above it is technically a fix.  There was an off-by-one array bug before.  However, I do not think that function is technically used either.

Bertram25 commented:
Overall change looks good AFAIK, my inline comment. Mind squashing your changes into a single commit after we've decided what to do about the game mode manager change. I think it will look cleaner. :)

authenticate commented:
Sure.  I'll update it.

authenticate commented:
It has been squashed. :)

Bertram25 commented:
Great! I'm testing now and I'll merge asap.
